Fix weird rustdoc output when single and glob reexport conflict on a name

GuillaumeGomez opened the pull request:
Fixes rust-lang/rust#143107.

The problem was that the second reexport would overwrite the first, leading to having unexpected results. To fix it, I now group items by their original `DefId` and their name and keep tracks of all imports for this item (should very rarely be more than one though, and even less often more than 2).

cc @lolbinarycat

GuillaumeGomez committed:
Fix weird rustdoc output when single and glob reexport conflict on a name

GuillaumeGomez committed:
Add regression test for #143107

lolbinarycat commented on src/librustdoc/visit_ast.rs:
````diff
@@ -32,11 +32,28 @@ pub(crate) struct Module<'hir> {
+    /// ```
+    ///
+    /// So in this case, we don't want to have two items but just one with attributes from both
+    /// imports to be merged.
````
is merging the attributes correct? i thought that [per the reference](https://doc.rust-lang.org/reference/items/use-declarations.html#r-items.use.glob.shadowing), only items not shadowed by non-glob imports would actually be shadowed.  i guess that doesn't exactly say what happens when you shadow a glob import with an identical item...

not that i see putting docs on glob imports to be a common usecase, i think much more common is docs on the regular import and no docs on the glob import.

GuillaumeGomez replied:
Oops, the comment is wrong: we never merge attributes from glob reexports.

GuillaumeGomez commented:
Ah bad luck. A function I modified was used in a more up-to-date version. ^^'

Anyway, rebased and fixed the CI failure.

lolbinarycat reviewed (commented):
some things could use clarification.  i don't entirely understand if things are getting merged or overwritten.  the unit test comment is confusing: all glob attrs are ignored, or just those on shadowed items? this should be clarified.

the test could also include coverage for the case where an item is renamed to itself.  also what happens if an item is renamed then renamed back to the original name in a chain?

GuillaumeGomez commented on src/librustdoc/clean/inline.rs:
```diff
@@ -152,8 +152,12 @@ pub(crate) fn try_inline(
     };
 
     cx.inlined.insert(did.into());
+    let import_def_ids: &[LocalDefId] = match import_def_id {
```
Wow, much better, thanks!

GuillaumeGomez commented:
> some things could use clarification. i don't entirely understand if things are getting merged or overwritten. the unit test comment is confusing: all glob attrs are ignored, or just those on shadowed items? this should be clarified.

Attributes on glob reexports are always ignored. The comment I added was misleading so I fixed it.

> the test could also include coverage for the case where an item is renamed to itself. also what happens if an item is renamed then renamed back to the original name in a chain?

What matters is the duo name+DefId. So if the name is different, then rustdoc considers it's a different item. And if the item as renamed as itself, then nothing happens either. But I'm gonna add a test for this second case just in case.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16147812381/job/45571327161) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45571327161) [(plain)](https://github.com/rust-lang/rust/commit/380ccc6f5ff701a33b09dac110e024e2fcd657d1/checks/45571327161/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:18
  local time: Tue Jul  8 15:44:31 UTC 2025
  network time: Tue, 08 Jul 2025 15:44:31 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lolbinarycat commented:
Looks like you rebased and responded to review in the the same force push, which was a bit confusing when I clicked on the "Compare" button and saw a bunch of unrelated changes.

Anyways, it looks mostly correct, although if nothing is getting merged in the unit test, maybe it shouldn't have `merge` in the name? Or maybe it should just also exercise the case where attributes are actually merge.. is there a case where attributes are actually merged?

Also, I think it would be nice to also test what shows on the module docs, since apparently that can get out of sync with the actual item's docs, as shown in the original issue.

GuillaumeGomez commented:
Ah I see. Added a check for short docs on module page.

GuillaumeGomez commented:
@bors r=lolbinary rollup

bors commented:
:pushpin: Commit 086b13d170b9207cc7d726048517a8cd5acc096d has been approved by `lolbinary`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lolbinary 086b13d170b9207cc7d726048517a8cd5acc096d -->
<!-- homu: {"type":"Approved","sha":"086b13d170b9207cc7d726048517a8cd5acc096d","approver":"lolbinary","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
